app/models/phylo_tree.rb

```diff
@@ -0,0 +1,147 @@
+class PhyloTree < ApplicationRecord
+  include PipelineOutputsHelper
+  include PipelineRunsHelper
+  has_and_belongs_to_many :pipeline_runs
+  belongs_to :user
+  belongs_to :project
+  validates :name, presence: true, uniqueness: true
+
+  STATUS_INITIALIZED = 0
+  STATUS_READY = 1
+  STATUS_FAILED = 2
+  STATUS_IN_PROGRESS = 3
+
+  def self.in_progress
+    where(status: STATUS_IN_PROGRESS)
+  end
+
+  def newick_s3_path
+    "#{phylo_tree_output_s3_path}/#{dag_version}/phylo_tree.newick"
+  end
+
+  def dag_version_file
+    "#{phylo_tree_output_s3_path}/#{PipelineRun::PIPELINE_VERSION_FILE}"
+  end
+
+  def monitor_results
+    # Retrieve dag version, which is needed to construct the output path:
+    update_pipeline_version(self, :dag_version, dag_version_file)
+    return if dag_version.blank?
+
+    # Retrieve output:
+    file = Tempfile.new
+    _stdout, _stderr, status = Open3.capture3("aws", "s3", "cp", newick_s3_path, file.path.to_s)
+    if status.exitstatus.zero?
+      file.open
+      self.newick = file.read
+      self.status = newick.present? ? STATUS_READY : STATUS_FAILED
+      save
+    end
+    file.close
+    file.unlink
+  end
+
+  def monitor_job(throttle = true)
+    # Detect if batch job has failed so we can stop polling for results.
+    # Also, populate job_log_id.
+    return if throttle && rand >= 0.1 # if throttling, do time-consuming aegea checks only 10% of the time
+    job_status, self.job_log_id, _job_hash, self.job_description = job_info(job_id, id)
+    if job_status == PipelineRunStage::STATUS_FAILED ||
+       (job_status == "SUCCEEDED" && !Open3.capture3("aws", "s3", "ls", newick_s3_path)[2].exitstatus.zero?)
+      self.status = STATUS_FAILED
+      save
+    end
+  end
+
+  def job_command
```

Comment:
On which thread is this executed?   On a webapp request response thread, or on the monitor main thread?

Reply:
If monitor, then we should make sure it doesn't take too long.    If webapp, I don't care.

Reply:
Neither -- it's executed as an asynchronous resque job enqueued upon the user's request. But the action responds immediately